v2/compiler/visitor.go

```diff
@@ -0,0 +1,135 @@
+package compiler
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/golang/protobuf/jsonpb"
+	"github.com/kubeflow/pipelines/api/v2alpha1/go/pipelinespec"
+)
+
+type Visitor interface {
+	Container(name string, component *pipelinespec.ComponentSpec, container *pipelinespec.PipelineDeploymentConfig_PipelineContainerSpec) error
+	Importer(name string, component *pipelinespec.ComponentSpec, importer *pipelinespec.PipelineDeploymentConfig_ImporterSpec) error
+	Resolver(name string, component *pipelinespec.ComponentSpec, resolver *pipelinespec.PipelineDeploymentConfig_ResolverSpec) error
+	DAG(name string, component *pipelinespec.ComponentSpec, dag *pipelinespec.DagSpec) error
+}
+
+const (
+	rootComponentName = "root"
+)
+
+func Accept(job *pipelinespec.PipelineJob, v Visitor) error {
+	if job == nil {
+		return nil
+	}
+	// TODO(Bobgy): reserve root as a keyword that cannot be user component names
+	spec, err := getPipelineSpec(job)
+	if err != nil {
+		return err
+	}
+	deploy, err := getDeploymentConfig(spec)
+	if err != nil {
+		return err
+	}
+	state := &pipelineDfsState{
+		spec:    spec,
+		deploy:  deploy,
+		visitor: v,
+		visited: make(map[string]bool),
+	}
+	return state.dfs(rootComponentName, spec.GetRoot())
+}
+
+type pipelineDfsState struct {
+	spec    *pipelinespec.PipelineSpec
+	deploy  *pipelinespec.PipelineDeploymentConfig
+	visitor Visitor
+	// records which DAG components are visited for circular reference detection
+	// TODO(Bobgy): fix circular reference detection for A -> B -> C & A -> C (there are no circles, but we will detect it now)
+	visited map[string]bool
+}
+
+func (state *pipelineDfsState) dfs(name string, component *pipelinespec.ComponentSpec) error {
+	if component == nil {
+		return nil
+	}
+	if state == nil {
+		return fmt.Errorf("dfs: unexpected value state=nil")
+	}
+	componentError := func(err error) error {
+		return fmt.Errorf("error processing component name=%q: %w", name, err)
+	}
+	executorLabel := component.GetExecutorLabel()
+	if executorLabel != "" {
```

Comment:
Is it valid when executorLabel is empty?

Reply:
Yes, if executor is not nil, then it's a DAG.